### ISSUE TITLE
DeepBGC integration

### ISSUE THREAD

**Issue by prihoda** [antismash/antismash#122]
Hi @kblin and team. We are happy about your involvement in the DeepBGC module, would you be interested in integrating the module in antiSMASH?

DeepBGC can be used programmatically, here's an example:

```python
from deepbgc import DeepBGCDetector, DeepBGCClassifier

os.environ["DEEPBGC_DOWNLOADS_DIR"] = os.path.join(os.path.basename(__file__), 'data')

detector = DeepBGCDetector('deepbgc', merge_max_protein_gap=1, merge_max_nucl_gap=2000, score_threshold=0.5)
product_classifier = DeepBGCClassifier('product_class')
activity_classifier = DeepBGCClassifier('product_activity')

detector.run(record)
product_classifier.run(record)
activity_classifier.run(record)
```

If you are interested, please let me know and I will be happy to get involved and discuss next steps.

**Comment by kblin** [comment]
Hi David,

We've discussed this previously (I had a quick chat with @Microbiology about this after the preprint was released). I think conceptually DeepBGC complements the rule-based nicely. Practically, I still see a number of challenges with integrating DeepBGC with antiSMASH. I'm happy to discuss those, though.

So I think in a recent update, you changed DeepBGC to also accept pre-annotated genomes, that does remove one of the issues I had on my list, thanks. That said, in antiSMASH we also run our own PFAM annotations. Since running full genome PFAM is the largest chunk of the runtime of DeepBGC and takes around 2-3 orders of magnitude longer than a regular antiSMASH run without PFAM annotations, I would like to be able to at least re-use these annotations for both tools.

Another issue I see is from a packaging perspective with the tensorflow library. We have seen things break in fun and entertaining ways between minor versions of the library and are not particularly fond of adding it on top of the already quite significant dependency stack. I'm aware that by design this is not something you can easily fix in DeepBGC, so one thing we discussed internally was if we could come up with a way to load DeepBGC annotations into antiSMASH, and then have antiSMASH also display the DeepBGC regions in the output. We don't currently have the code to do this, but this would be trivial to implement.

I'm flexible as to how we connect things up from a UI perspective on the submission website. We could have a separate website for running DeepBGC that then allows passing on the results to antiSMASH, or support this directly from the antiSMASH submission page, and then internally have a processing step that first runs DeepBGC and then runs antiSMASH.

Does that sound like a solution you might be interested in?

**Comment by prihoda** [comment]
Hi Kai, 

that's great to hear. I think we can definitely address those technical challenges. 

Concerning the dependencies, would you feel comfortable about adding DeepBGC as an optional dependency? The DeepBGC features would be available only if the user installs the additional deepbgc module in the same environment. I have tested that approach on Mac and linux machines. But if you feel like it could introduce incompatibility issues, running DeepBGC separately and loading the result in antiSMASH seems like a reasonable solution.

As for the PFAM annotations, I will implement support for the antiSMASH PFAM_domain feature format (using "db_xref"). One thing to discuss is that DeepBGC requires PFAM_domain features to have a "database" qualifier that has to be equal to "Pfam-A.31.hmm". This is done in order to guarantee that we are using the same version as was used for training the model. We could expect the model to work on other versions of the database but we have not performed any evaluation, therefore we felt safer requiring a specific version. Do you think this "database" qualifier could be added also by antiSMASH or would you rather add a DeepBGC parameter that disables this strict requirement?

Thank you,
David

**Comment by SJShaw** [comment]
antiSMASH already annotates PFAM_domain features with a `database` qualifier that stores the full Pfam-A database version used for detection: e.g. `/database="31.0"`.

**Comment by prihoda** [comment]
Hi everyone, I updated the PFAM_domain annotations to be compatible: https://github.com/Merck/deepbgc/pull/6 

Can you please try to download the current 0.1.5 DeepBGC version from Bioconda or PyPI and let me know how I can help with the next steps?

**Comment by SJShaw** [comment]
I've tried out 0.1.5, since that was what was available with pip when I installed it yesterday (and I don't see any changes between that and 0.1.6 that would affect what much of what follows). There's a few issues that I can see that might take a while to work out. We'll be releasing the final 5.0 within the next two weeks (since the manuscript has already been provisionally accepted), so our flexibility for changing core parts of our infrastructure is limited.

The main issue I see with some kind of integration is that we can't use your clusters as we use clusters in antiSMASH 5:
- aS5 clusters must have one, and only one, product, where you can have no product prediction and multiple products predicted
- what data is expected of a cluster differs too much (while this can be changed in antiSMASH, that's too much work to put into the 5.0 release that will be happening in the next two weeks)
- product names vary too much to allow product-specific analyses to run, e.g. your `RiPP` compared to our more specific versions `lanthipeptide`, etc which each have their own analysis chain

Other issues (for programmatic access and subprocess-based access):
- `scikit-learn` versions differ, which may cause problems with unpickling trained datasets, we use 0.19, you pickled with 0.18.3 and install with the latest version (0.20.3 as of yesterday), fixing this would require us keeping our `scikit-learn` versions in sync and retraining to update
- we can't provide the records as we use them as your input, we use a customised access layer (see `antismash.common.secmet`), they'd first have to be converted to biopython, and checking the output still requires accessing output files to get only what changed
- dependency issues as mentioned by Kai

Kai mentioned the option of loading annotations as part of antiSMASH input. We can use your cluster predictions like we use the clusterfinder predictions (see `antismash.common.secmet.features.SubRegion`). These do not affect product-specific analyses, but can be used with the clusterblast module for comparison to clusters in the antiSMASH and MIBiG databases. It's quite easy to implement something like this, but we'd need to work out what kind of information is to be kept beyond location and how it should be stored, especially if it's to be generic to support other tools. I'd expect it to be some kind of JSON or simple tab separated format, depending on how much information is to be kept. For antiSMASH purposes, at a minimum it would need to include:
- record identifier as given in the input files to antismash
- location (start and end, strand would be irrelevant)
- name of the tool making the predictions
- some kind of single description of the prediction (e.g. your product)
There's other information that could be useful for users, like tool version, scores, etc, and that's what we'd need to decide on.

Practically speaking, I think the side loading of annotations is the best option to go for.

**Comment by prihoda** [comment]
Thanks for the detailed analysis. I will try to think about all the points. As for loading annotations, DeepBGC provides all of these properties in the bgc.tsv file (see example.bgc.tsv in https://nbviewer.jupyter.org/urls/github.com/Merck/deepbgc/releases/download/v0.1.0/DeepBGC_Example_Result.ipynb):

- record identifier: `sequence_id`
- location: `nucl_start`, `nucl_end`
- name of the tool making the predictions: `detector_label`
- some kind of single description of the prediction: `product_class` (could also be just 'DeepBGC Putative' or if a longer text description is accepted, an aggregation of multiple columns)

**Comment by SJShaw** [comment]
The issue with your current output (from the point of view of antiSMASH parsing it) is that if you change that table in any way between versions of DeepBGC (which seems to have happened between version 0.1.0 and 0.1.5 already), it can easily cause problems because it's a flat file and isn't embedding any information about version (and even if it did embed version, then different versions would require a parser for each output version). We've had problems in the past trying to keep contributed code in sync for external tools, and we don't want to repeat that experience.

A well defined and simply parsed data format is required for this import/sideload of annotations. A JSON format would allow us to set a standard that will work with a variety of inputs (and is easily parsed and written with dictionaries and the standard python `json` library). It would allow you to add different values easily (e.g. if you add a new classification type with another score). Something like this perhaps:
```
{
  "name": "deepbgc",
  "version": 1.2.3,
  // possible also storing something about the run configuration, like minimum thresholds
  "records": [
      "recordName": "first_record_id",
      "areas": [ // with one item in the array for each of your cluster predictions
          {
            "start": 158,
            "end": 2510,
            "label": "Polykeytide", //or something else brief
            ... // other mandatory fields or optional fields that may be used by antiSMASH in some way
            "details": { // embeddable as extra qualifiers in genbank output, could be empty
              "score": 6.5,
              "some_option_name": true,
              ... // as long as it's valid for biopython qualifiers, this could store as much as you like
            }
          },
          ... // continuing with other predictions, if any
        ],
    ...  // continuing with predictions for other records, if any
    ]
}
```
What information is stored in such a data format and where it is stored is that part that we need to define. For example, we couldn't add all the information from your table output to the antiSMASH HTML output, but it would be useful to add more than 'DeepBGC Putative', since there are (usually) more useful details available.

To help with working out such a schema, here are some questions:
- Which details would you like to have visible in antiSMASH HTML output?
- Which details would you like to have annotated in the antiSMASH genbank output?
- Are you planning to add annotations in the future that would be included in either of the above?

**Comment by prihoda** [comment]
I see, a JSON format with mandatory fields and a dictionary of optional fields would definitely help here. The only problem I see with the JSON format is that per-pfam and per-protein scores cannot be provided, but that is definitely not our top priority.

- In the HTML output we would like to have visible for each cluster: 
  - detector name (name of trained detection model, e.g. `deepbgc`, more model versions might appear in the future, e.g. `deepbgc-mibig1.4-full`, `deepbgc-mibig1.4-curated`)
  - detector label (an optional human-readable user-defined detection label, e.g. DeepBGC strict)
  - location
  - label = predicted product classes. Or alternatively we could provide only the single top-scoring class, though that will not respect the multi-label architecture of the model.
  - score
  - cluster & detector details - dictionary of optional attribute-value pairs (if possible, shown e.g. as bullet-points on click)
- In the GenBank output we would like to have visible the same cluster fields as mentioned above. The detector metadata could be stored as a structured comment in each record
- Currently we do not have any plans to add additional annotations.

How would you feel about the following output?

```
{
  "module": "deepbgc",
  "module_version": "0.2.3", // version of deepbgc module used to generate this file
  "name": "deepbgc", 
  "label": "DeepBGC strict",
  "details": { 
    "detector_version": "0.1.0", // version of trained model (will be less or equal to deepbgc module version)
    "score_threshold": 0.9,
    "merge_max_nucl_gap": 0,
    "merge_max_protein_gap": 0,
    "min_proteins": 1,
    "min_domains": 1,
    "min_bio_domains": 0
  },
  "records": [
    {  
      "recordName": "first_record_id",
      "areas": [ // with one item in the array for each cluster prediction
        {
          "start": 158,
          "end": 2510,
          "label": "Polykeytide,NRP",
          "score": 0.94, 
          "details": { 
            "bgc_candidate_id": "GG657747.1_30710-39138.1",
            "product_class": "Polykeytide,NRP",
            "product_class_score": "Alkaloid=0.00,NRP=0.14,Other=0.20,Polyketide=0.26,RiPP=0.37,Saccharide=0.03,Terpene=0.05",
            "product_activity": "",
            "product_activity_score": "antibacterial=0.46,cytotoxic=0.36,inhibitor=0.21,antifungal=0.08"
          }
        },
        // ...
        // continuing with other predictions, if any
      ],
    }
    // ...
    // continuing with predictions for other records, if any
  ]
}
```

DeepBGC can be executed multiple times on the same record with different detection parameters. Each detection is identified by a unique user-defined "detector label" (e.g. DeepBGC strict, DeepBGC lenient). These detections would be provided as separate files. But it would also be sufficient to accept just one detector output, since that is the most common usecase.

**Comment by SJShaw** [comment]
For the JSON format, we'd like to make this a generic standard for importing other annotations into antiSMASH, so we'll take a little time to ensure other use cases we can think of will be handled.

For the HTML output, I'm not sure how to add that much information without causing some interaction issues (or just overloading users). For a demo, I hacked in a very basic version of sideloading to put in your analysis results of NC_003888, with the results available [here](https://antismash.secondarymetabolites.org/upload/NC_003888_with_deepbgc/), (ignore the CASSIS part of the legend, it's just a result of how I hacked this in). It doesn't include anything other than product and location (and uses `unknown` when no product predicted). If you can create a simple mock-up of what you'd expect it to look like/behave, that would be helpful.

**Comment by prihoda** [comment]
Such output would already be very useful. Additional information could be shown in a generic "Region details tab" using a table:

<img width="900" alt="Screen Shot 2019-03-27 at 15 20 46" src="https://user-images.githubusercontent.com/2894124/55084770-c6675f00-50a5-11e9-9d90-55721a186b32.png">

It would be great to be able to color-code the results based on the product type predictions, but I understand that it would have to map to a common list of product types. If we train the Random Forest classifier on the antiSMASH classification of the MIBiG dataset, would you consider showing the classes using your color scheme? The goal of our machine-learning approach is that users can easily train their own detection or classification models simply using `deepbgc train`.

Predicted products could be shown using border color only:

<img width="766" alt="Screen Shot 2019-03-27 at 15 44 30" src="https://user-images.githubusercontent.com/2894124/55085627-42ae7200-50a7-11e9-8e34-098173fb6f6d.png">

**Comment by SJShaw** [comment]
Tabs will be reasonably easy to create from details. Your mockup covers just the DeepBGC run options, but I assume you'd also want the details of the prediction present as well.
Also, since multiple DeepBGC subregions can land inside a single region (e.g. [here](https://antismash.secondarymetabolites.org/upload/NC_003888_with_deepbgc/#r1c126)) we can have a single tab per DeepBGC subregion or list them all in one tab fairly easily. I'd probably lean to one tab per subregion and have the tool run options/config in an expandable section since if a user has seen it once they don't need to see it every time.

For the colours, this will get a little complicated. There's a few things to consider here:
- For our cluster types/classifications, the colouring also guarantees certain information, e.g. A-domain specificity analyses for NRPS cluster candidates (the green ones). If you retrain and we recolour, this would weaken that link between type and analysis.
- If users can train their own classifications, there's no guarantee that this side-loading will result in matching classifications anyway, so retraining the defaults to ours simply for colouring is probably not worthwhile (I could be wrong here about how likely it is for people to retrain). What classifications you use is probably best determined by what gives the best prediction results (e.g. your RiPP vs our specific RiPP subclasses, if your predictions for subclasses are worse, it's best not to coerce them just for display reasons).
- I think we're already a bit too noisy with colours, and while the border/shadow option you have works well for some colours, the contrast problems (e.g. grey/white combinations) might end up being too much. We also would need a case for what to do with multiple predictions in the same region, like our current hybrids.

Going back to the JSON format, I haven't thought of any serious issues with it. Apart from a few name changes for clarity (like `configuration` instead of details for the tool in general, perhaps), and changing the per-region details to contain lists (for nicer conversion and formatting as genbank qualifiers), I think it'll be about the as proposed. I'll start on a branch for this with a JSON schema that we can validate inputs against, then we can try it out some real DeepBGC outputs and see if we find any other issues.

**Comment by SJShaw** [comment]
@prihoda, I've just submitted a draft pull request that should have all the required functionality for adding your annotations. It doesn't have any special colouring as of the moment.

An example input JSON would look like [this](https://github.com/antismash/antismash/blob/138c3358c9127306b8781ed0de5126b4452df626/antismash/detection/sideloader/test/data/good.json).

If you've going any feedback about it, I'd love to hear it. If you've come across some other use cases, I'm happy to try and work them in too.

**Comment by prihoda** [comment]
I created a DeepBGC branch that produces the JSON output: https://github.com/Merck/deepbgc/tree/antismash 

At the moment it only creates `subregions`, not `protoclusters` since our classification is not aligned. 

Here is an example JSON file and its corresponding GBK output file:
[Streptomyces_pristinaespiralis.zip](https://github.com/antismash/antismash/files/3286203/Streptomyces_pristinaespiralis.zip)

DeepBGC can run multiple models (DeepBGC, ClusterFinder retrained, ClusterFinder GeneBorder) and it also preserves clusters that were manually annotated before the sequence was processed with DeepBGC. At the moment, all detected clusters are stored in a single JSON file with a DeepBGC tool header. The specific detector information is stored in each region's details. In the final version one JSON file should be created for each detector.

I think at this stage, this is already a valuable feature. Would you consider starting a discussion about running DeepBGC from the antiSMASH submission page, as you mentioned in your first response? Unfortunately we don't have access to any public-facing servers that would handle such high computation load.

**Comment by SJShaw** [comment]
The JSON example you have seems fine. It has shown there's a couple of issues with my draft implementation, which is handy.

For the content of your JSON, the only thing I'd recommend changing is the "label" field. You're using `"Putative BGC"` for every label, which is the main piece of information being shown to a user in our HTML output. It might be a better idea to use your `product_class` or `product_activity` fields (possibly also including whether it came from annotations or the machine learning portion). The region labels in the HTML would look like `DeepBGC #X: Polyketide` instead of all `DeepBGC #X: Putative BGC`. This is entirely up to you though. All the information given in the JSON will still be available in the detail sections of the HTML.

As for running DeepBGC via the antiSMASH submission page, it's separate from the implementation parts, so I'll let Kai answer that part.

**Comment by kblin** [comment]
Capacity is a bit of an issue on our side as well. The full genome PFAM annotations are the time intensive part of the analysis, and we simply can't afford to run that for every job getting submitted.

A quick back of the envelope calculation here:
* It takes us roughly 20 minutes to run antiSMASH on a ~10 Mbp genome with all the features enabled.
* A very optimistic estimate for getting a full-genome PFAM annotations for that same genome is at least 2 hours.
* We are now running something like 500 jobs a day, at 0.33 hours per job that's ~ 167 "compute hours"
* Adding PFAM annotations adds another 1000 compute hours on top of that, meaning we'd require at least 10x the capacity we now provide

Now of course you could argue that not everybody would be running DeepBGC, but our experience with providing ClusterFinder is that users like to enable all the features the submission website has to offer, and we were running over half of our jobs with ClusterFinder enabled. So even at this conservative estimate, we're looking at 5x the capacity we run now.

That being said, I'd be happy to provide a DeepBGC service on the secondarymetabolites.org infrastructure. I would like to keep it a bit more separate, so it will only get used by people who explicitly want to run DeepBGC.

I'll also be at the ISMB/ECCB for the OBF-BOSC COSI, I'm happy to meet up there to discuss this in more detail.

**Comment by SJShaw** [comment]
Now that sideloading of deepBGC results is possible and there haven't been any issues pop up for it, I'll close this for now.
